clusterchecks: simple dispatching logic

xvello opened the pull request:
### What does this PR do?

Add a simple dispatching logic for cluster-checks to account for node-agent churn:

- configs are assigned to the node that has the less configs assigned to it
- nodes are removed from the store if they don't report for 30 seconds (configurable)
- configs dispatched to expired nodes are re-dispatched on other available nodes
- if no nodes are available, configs are dispatched to a dummy node with an empty hostname. It will expire in 15 seconds and we will try re-dispatching the configurations

As the lifecycle of the package will become more complex when adding the leader election handling, I'm experimenting with the use of contexts to handle the sub-routines. If that does not prove better, I'll move back to stop channels.

### Testing

Logs of a test session with three clients:

- Start the cluster-agent with no node-agent: config is in standby
- Start `node1` and `node2`, config gets delegated to `node1`
- Stop `node1`, it expires and config is delegated to `node2`
- Stop `node2`, no more available nodes, config is in standby
- Start `node3`, config gets delegated to it

> 2018-10-01 16:24:37 UTC | INFO | (dispatcher_main.go:62 in add) | No available node to dispatch http_check:22f5b2455bae3d2d on, will retry later
> 2018-10-01 16:24:52 UTC | INFO | (dispatcher_main.go:64 in add) | Dispatching configuration http_check:22f5b2455bae3d2d to node node1
> 2018-10-01 16:25:52 UTC | INFO | (dispatcher_nodes.go:84 in expireNodes) | expiring out node node1, last status report 35 seconds ago
> 2018-10-01 16:25:52 UTC | INFO | (dispatcher_main.go:64 in add) | Dispatching configuration http_check:22f5b2455bae3d2d to node node2
> 2018-10-01 16:27:07 UTC | INFO | (dispatcher_nodes.go:84 in expireNodes) | Expiring out node node2, last status report 37 seconds ago
> 2018-10-01 16:27:07 UTC | INFO | (dispatcher_main.go:62 in add) | No available node to dispatch http_check:22f5b2455bae3d2d on, will retry later
> 2018-10-01 16:27:22 UTC | INFO | (dispatcher_main.go:62 in add) | No available node to dispatch http_check:22f5b2455bae3d2d on, will retry later
> 2018-10-01 16:27:37 UTC | INFO | (dispatcher_main.go:62 in add) | No available node to dispatch http_check:22f5b2455bae3d2d on, will retry later
> 2018-10-01 16:27:52 UTC | INFO | (dispatcher_main.go:62 in add) | No available node to dispatch http_check:22f5b2455bae3d2d on, will retry later
> 2018-10-01 16:28:07 UTC | INFO | (dispatcher_main.go:64 in add) | Dispatching configuration http_check:22f5b2455bae3d2d to node node3

xvello committed:
clusterchecks: simple dispatching logic

CharlyF commented on pkg/clusteragent/clusterchecks/handler.go:
```diff
@@ -23,49 +24,48 @@ type Handler struct {
 }
 
 // SetupHandler returns a populated Handler
 // If will hook on the specified AutoConfig instance at Start
```
s/If/It

xvello committed:
s/SetupHandler/NewHandler/

xvello commented on pkg/clusteragent/clusterchecks/dispatcher_main.go:
```diff
@@ -48,11 +54,17 @@ func (d *dispatcher) add(config integration.Config) {
+	if target == "" {
+		// If no node is found, dispatch it to the dummy "" node.
+		// The node will be expired out, and configurations dispatched on a real node later.
+		log.Infof("No available node to dispatch %s:%s on, will retry later", config.Name, config.Digest())
```
That's a pretty bad situation we're in if we're printing this: we have checks to run but nowhere to run them. I think it's worth an info line as we're not doing what's expected of us.
Hopefully, this should not happen on a prod cluster, unless the whole daemonset is down, or the network has issues

CharlyF replied:
Oh I see! I concur

hkaj replied:
then maybe a warning?

xvello committed:
pr feedback

CharlyF commented on pkg/clusteragent/clusterchecks/dispatcher_nodes.go:
```diff
@@ -43,3 +44,53 @@ func (d *dispatcher) processNodeStatus(nodeName string, status types.NodeStatus)
+// expireNodes iterates over nodes and removes the ones that have not
+// reported for more than the expiration duration. It returns configs
+// that were dispatched to these nodes, to re-dispatch them.
+func (d *dispatcher) expireNodes() []integration.Config {
```
I think the name of this method only partially covers its scope.
It does delete the stale nodes from the store  but it also returns the orphaned configs.
Could we have something like

```
func (d *dispatcher) retrieveStaleState() ([]integration.Config, staleNodes string) {
...
~delete(d.store.nodes, name)~
staleNodes = append(staleNodes, name)
...
return orphanedConfigs, staleNodes
```
And delete the nodes there https://github.com/DataDog/datadog-agent/pull/2392/files#diff-bd226a7b7bfd2b2f5a425bae0935f3cdR92 ?

xvello replied:
That would indeed be more readable, but would not be atomic anymore.
Doing that, we might run into a race condition where we would schedule a config on an expired node between the retrieve and delete calls. We'd lose it, as it would not be listed in `orphanedConfigs`

CharlyF commented on pkg/clusteragent/clusterchecks/dispatcher_main.go:
```diff
@@ -61,6 +73,23 @@ func (d *dispatcher) remove(config integration.Config) {
+		case <-ctx.Done():
+			return
+		case <-expireTicker.C:
+			expired := d.expireNodes()
```
I would rename expired to orphanedConfigs (so that it's not ambiguous with the expireNode method)

CharlyF commented on pkg/clusteragent/clusterchecks/dispatcher_main.go:
```diff
@@ -48,11 +54,17 @@ func (d *dispatcher) add(config integration.Config) {
-	// TODO: add dispatching logic
-	hostname, _ := util.GetHostname()
-	d.addConfig(config, hostname)
+	target := d.getLeastBusyNode()
```
What do you think about using a queue to store the pending configs to dispatch instead of scheduling them to the dummy node ?
This can be for later, just wanted your thoughts on this ?

(still trying to figure out if the dummy node can introduce latency on the scheduling if we wait for it to be stale)

xvello replied:
That would be cleaner indeed, although it's basically a safety net if we loose all agents than a common code path.
Max latency DCA-side will be by 30/2 = 15 seconds by default. Add 10 seconds for the node-agent pool frequency

xvello replied:
Implemented this: dangling configs are now stored in a slice, that is retried every 15 seconds: https://github.com/DataDog/datadog-agent/pull/2392/files#diff-62d6072bb2f0d7d53f6f4299e5395fe9R30

xvello committed:
compile in clustercheck build tag by default

xvello committed:
rename orphanedConfigs

xvello committed:
use the danglingConfigs maps instead of a dummy "" node

xvello committed:
rework expireNodes scope

hkaj commented on pkg/config/config.go:
```diff
@@ -330,6 +330,7 @@ func initConfig(config Config) {
 	config.BindEnvAndSetDefault("external_metrics_provider.local_copy_refresh_rate", 30) // value in seconds
 	// Cluster check Autodiscovery
 	config.BindEnvAndSetDefault("cluster_checks.enabled", false)
+	config.BindEnvAndSetDefault("cluster_checks.node_expiration_timeout", 30) // value in seconds
```
could we add these settings to the [config template](https://github.com/DataDog/datadog-agent/blob/master/pkg/config/config_template.yaml) with descriptions?

xvello replied:
Done

hkaj commented on pkg/clusteragent/clusterchecks/dispatcher_main.go:
```diff
@@ -48,11 +54,16 @@ func (d *dispatcher) add(config integration.Config) {
+		// If no node is found, store it in the danglingConfigs map for retrying later.
+		log.Infof("No available node to dispatch %s:%s on, will retry later", config.Name, config.Digest())
+	} else {
+		log.Infof("Dispatching configuration %s:%s to node %s", config.Name, config.Digest(), target)
```
i'm on the fence between info and debug for this one. I guess info is fine but we may end up lowering to debug if this runs smoothly and it gets spammy.

xvello replied:
I went for consistency with [the collector's scheduler](https://github.com/DataDog/datadog-agent/blob/27e6bb2d015681f25850ed4376f41f57350d19da/pkg/collector/scheduler.go#L63), but indeed on high service churn that might become spammy.

CharlyF reviewed (dismissed):
Love the testing suite.
🐐

hkaj commented on pkg/clusteragent/clusterchecks/dispatcher_nodes.go:
```diff
@@ -43,3 +44,50 @@ func (d *dispatcher) processNodeStatus(nodeName string, status types.NodeStatus)
+
+	for name, node := range d.store.nodes {
+		node.RLock()
+		if node.lastPing < cutoffTimestamp {
```
I didn't catch that earlier, but can we talk briefly about the `lastPing`?
If I understand correctly we use the ping from the client (that checks connection with the cluster agent, and triggers the polling of new configs) as a heartbeat. Is that correct? If so we should explain that somewhere, and maybe rename lastPing in the DCA code to heartbeat. lastPing is fine client side but heartbeat would work as well, your call.
Let me know if I missed something.

hkaj commented on pkg/clusteragent/clusterchecks/stores.go:
```diff
@@ -20,16 +20,18 @@ import (
+	digestToConfig  map[string]integration.Config // All configurations to dispatch
+	digestToNode    map[string]string             // Node running a config
+	nodes           map[string]*nodeStore         // All nodes known to the cluster-agent
+	danglingConfigs map[string]integration.Config // Configs we count not dispatch to any node
```
count --> could

hkaj reviewed (changes requested):
Nice PR, I like that it's bitesize and well integrated to the existing, still with some minor revamp improvements 💯
Left a few comments but looks good overall

xvello committed:
address PR feedback

- update config template
- rename lastPing > heartbeat and document in readme
- move "no node" message to warning

hkaj reviewed (approved):
⭐️
